File: src/main/java/io/github/tanguygab/conditionalactions/actions/items/TakeItemAction.java
package io.github.tanguygab.conditionalactions.actions.items;

import io.github.tanguygab.conditionalactions.actions.Action;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TakeItemAction extends Action {

    public TakeItemAction() {
        super("(i?)take-item:( )?");
    }

    @Override
    public String getSuggestion() {
        return "take-item: <material> <amount> <name> <lore>";
    }

    @Override
    public void execute(OfflinePlayer player, String match) {
        if (!(player instanceof Player p)) return;

        String[] args = match.split(" ");
        String material = args[0].toUpperCase();

        Material mat = Material.getMaterial(material);
        if (mat == null) return;

        int amt = args.length > 1 ? parseInt(parsePlaceholders(p,args[1]),1) : 99999;
        if (amt <= 0) return;

        String name = null;
        List<String> lore = new ArrayList<>();

        if (args.length > 2) {
            args = Arrays.copyOfRange(args,2,args.length);
            args = String.join(" ",args).split("\\n");
            name = parsePlaceholders(p,args[0]);

            if (args.length > 1)
                for (String line : Arrays.copyOfRange(args,1,args.length))
                    lore.add(parsePlaceholders(p, line));
        }

        PlayerInventory inv = p.getInventory();
        ItemStack[] items = inv.getContents();
        for (int i = 0; i < items.length; i++) {
            if (amt <= 0) break;
            if (!check(items[i],mat,name,lore)) continue;
            ItemStack found = items[i];
            if (found.getAmount() <= amt) {
                inv.setItem(i,null);
                amt = amt-found.getAmount();
                continue;
            }
            if (found.getAmount() > amt) {
                found.setAmount(found.getAmount()-amt);
                amt = 0;
            }
        }
    }

    public boolean check(ItemStack item, Material mat, String name, List<String> lore) {
        if (item == null || item.getType() != mat) return false;
        ItemMeta meta = item.getItemMeta();
        assert meta != null;
        if (name != null && !meta.getDisplayName().equals(name)) return false;
        if (!lore.isEmpty()) {
            if (!meta.hasLore()) return false;
            List<String> itemLore = meta.getLore();
            if (itemLore == null) return false;
            if (lore.size() != itemLore.size()) return false;

            for (String line : itemLore) {
                int pos = itemLore.indexOf(line);
                if (!lore.get(pos).equals(line)) return false;
            }
        }

        return true;
    }

}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/actions/types/items/TakeItemAction.kt
package io.github.tanguygab.conditionalactions.actions.types.items

import io.github.tanguygab.conditionalactions.actions.Action
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer
import org.bukkit.Material
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

class TakeItemAction : Action("^(i?)take-item:( )?".toRegex()) {
    private val mmLegacy get() = LegacyComponentSerializer.legacyAmpersand()

    override fun getSuggestion() = "take-item: <material> <amount> <name> <lore>"

    override fun execute(player: OfflinePlayer?, match: String) {
        if (player !is Player) return

        var args = match.split(" ")
        val material = args[0].uppercase()

        val mat = Material.getMaterial(material) ?: return

        var amt = if (args.size > 1) parseInt(parsePlaceholders(player, args[1]), 1) else 99999
        if (amt <= 0) return

        var name: String? = null
        val lore = mutableListOf<String>()

        if (args.size > 2) {
            args = args.subList(2, args.size).joinToString(" ").split("\\n")
            name = parsePlaceholders(player, args[0])

            if (args.size > 1) for (line in args.subList(1, args.size))
                lore.add(parsePlaceholders(player, line))
        }

        val inv = player.inventory
        val items = inv.contents
        for (i in items.indices) {
            if (amt <= 0) break
            val found = items[i]
            if (found == null || !check(found, mat, name, lore)) continue

            if (found.amount <= amt) {
                inv.setItem(i, null)
                amt -= found.amount
                continue
            }
            if (found.amount > amt) {
                found.amount -= amt
                amt = 0
            }
        }
    }

    fun check(item: ItemStack, mat: Material, name: String?, lore: List<String>): Boolean {
        if (item.type !== mat) return false
        val meta = item.itemMeta
        if (name != null && meta != null && mmLegacy.serialize(meta.displayName() ?: Component.empty()) != name) return false
        if (lore.isNotEmpty() && meta != null) {
            if (!meta.hasLore()) return false
            val itemLore = meta.lore() ?: return false
            if (lore.size != itemLore.size) return false

            for (line in itemLore) {
                val pos = itemLore.indexOf(line)
                if (lore[pos] != mmLegacy.serialize(line)) return false
            }
        }

        return true
    }
}
